### PR TITLE
TLS: Improve server version negotiation testing

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Description
Improve the testing of the version negotiation on server side.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** not required, testing only
- [x] **backport** not required, no TLS 1.3 support in 2.28
- [x] **tests** not required

### CODE REVIEW

**Comment by tom-cosgrove-arm** [comment]
@ronald-cron-arm is this PR ready for review, or are you still working on it?

**Comment by tom-cosgrove-arm** [comment]
Tagging this as `very-high` as we want it in the release

**Comment by ronald-cron-arm** [comment]
The OpenCI PR tests has run successfully, the ABI-API check failure is not related to this PR, @tom-cosgrove-arm @yanesca this is ready for review.

**Review by mpg — APPROVED** [review]
LGTM, thanks!

I have a couple suggestions and questions, but none of them is important enough to delay this PR.

Note: I'd also checked that the new test cases added are indeed executed at least once in the CI (as opposed to being skipped in all tested configs).

**Review by yanesca — APPROVED** [review]
Looks good to me.

The tests cover all the meaningful combinations I could think of. I constructed tables like this to verify:
```
## Side: Server

Compile time    Runtime P   Runtime G
1.2             -           1.2
1.2, 1.3        -           1.2
1.2, 1.3        1.2         1.2

1.3             -           1.3
1.2, 1.3        -           1.3
1.2, 1.3        1.3         1.3

1.2, 1.3        -           all
1.2, 1.3        1.2         all
1.2, 1.3        1.3         all
1.3             -           all
1.2             -           all

all             -           1.0
all             -           1.1

1.3             -           1.2
1.2             -           1.3
1.2, 1.3        1.3         1.2
1.2, 1.3        1.2         1.3
```

When doing so, I ignored the title lines, because I found it easier to look at the actual parameters than decipher the title. I think Manuel's suggestions would make the test title easier to read, so +1 on those.

File: tests/ssl-opt.sh [hunk 1]
```diff
@@ -6930,37 +6922,30 @@ run_test    "Server selecting TLS 1.3, over TLS 1.2 if supported" \
...
-requires_config_enabled MBEDTLS_SSL_SRV_C
-requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
-requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3
-requires_config_enabled MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
-run_test    "Server selecting TLS 1.2 over TLS 1.3" \
-            "$P_SRV crt_file=data_files/server5.crt key_file=data_files/server5.key" \
+requires_all_configs_enabled MBEDTLS_SSL_SRV_C \
+                             MBEDTLS_SSL_PROTO_TLS1_2 MBEDTLS_SSL_PROTO_TLS1_3 \
+                             MBEDTLS_SSL_TLS1_3_COMPATIBILITY_MODE
+run_test    "Server version nego check G->m: [1.2]+1.3 / 1.2+1.3 -> 1.2" \
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply tests/ssl-opt.sh:7081]: It is to outline that 1.2 with this way of configuring GnuTLS is the preferred one.

**Pull request merged**